rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,549 @@
+- Signing of agent artifacts via TUF
+- Teleport Cloud APIs for updating agents
+- Improvements to the local functionality of the Kubernetes agent for better compatibility with FluxCD and ArgoCD.
+
+This RFD proposes a specific implementation of several sections in https://github.com/gravitational/teleport/pull/39217.
+
+Additionally, this RFD parallels the auto-update functionality for client tools proposed in https://github.com/gravitational/teleport/pull/39805.
+
+## Why
+
+The existing mechanism for automatic agent updates does not provide a hands-off experience for all Teleport users.
+
+1. The use of system package management leads to interactions with `apt upgrade`, `yum upgrade`, etc. that can result in unintentional upgrades.
+2. The use of system package management requires complex logic for each target distribution.
+3. The installation mechanism requires 4-5 commands, includes manually installing multiple packages, and varies depending on your version and edition of Teleport.
+4. The use of bash to implement the updater makes changes difficult and prone to error.
+5. The existing auto-updater has limited automated testing.
+6. The use of GPG keys in system package managers has key management implications that we would prefer to solve with TUF in the future.
+7. The desired agent version cannot be set via Teleport's operator-targeted CLI (tctl).
+8. The rollout plan for the new agent version is not fully-configurable using tctl.
+9. Agent installation logic is spread between the auto-updater script, install script, auto-discovery script, and documentation.
+10. Teleport contains logic that is specific to Teleport Cloud upgrade workflows.
+11. The existing auto-updater is not self-updating.
+12. It is difficult and undocumented to automate agent upgrades with custom automation (e.g., with JamF). 
+
+We must provide a seamless, hands-off experience for auto-updates that is easy to maintain.
+
+## Details - Teleport API
+
+Teleport will be updated to serve the desired agent version and edition from `/v1/webapi/find`.
+The version and edition served from that endpoint will be configured using the `cluster_maintenance_config` and `autoupdate_version` resources.
+Whether the updater querying the endpoint is instructed to upgrade (via `agent_auto_update`) is dependent on the `host=[uuid]` parameter sent to `/v1/webapi/find`.
+
+To ensure that the updater is always able to retrieve the desired version, instructions to the updater are delivered via unauthenticated requests to the `/v1/webapi/find`.
+Teleport proxies use their access to heartbeat data to drive the rollout and modulate the `/v1/webapi/find` response given the host UUID.
+
+Rollouts are specified as interdependent groups of hosts, selected by resource label.
+A host is eligible to upgrade if the label is present on any of its connected resources.
+
+At the start of a group rollout, the Teleport proxy marks a desired group of hosts to update in the backend.
```

Comment:
After chatting with @espadolini, I'm going to move the write logic that progresses the rollout to Auth.

Let me know if this looks right, and I'll add it to the RFD:

For each window, Auth will mark the matching instance heartbeats as eligible, and the eligible list will not change after the window starts. Additionally, `max_in_flight` instance heartbeats will be marked as upgrading, which causes the proxy to serve `auto_upgrade: true` from its cache of instance heartbeats. The "upgrading" state is stored with the instance heartbeat as a timestamp capturing the time the instance transitioned to upgrading. Every time an eligible instance heartbeat is updated by an incoming heartbeat, the heartbeats are evaluated to determine if additional heartbeats need to be marked as upgrading. This evaluation will be based on whether the existing eligible heartbeats are still upgrading, timed-out, or failed, using the upgrade timestamps.

The eligibility marker can also be a timestamp (along with group name), which expires after the window. The whole process requires two writes, which should be insignificant additional load.

The model is designed to be similar to how Kubernetes progresses replicasets with eventually-consistent reconciliation.

Let me know what you think 🙂

Reply:
> There is currently no communication between Auth servers, so coordinating between them which instance heartbeats are marked as eligible while staying below max_in_flight is not going to work.

I think that's okay. If it's difficult to make eligibility atomically or implement leader/follower, then instead of marking the initial set as eligible, we can drive the entire process via heartbeat writes.

> Neither the Auth or Proxy currently cache instance heartbeats.

Would you suggest adding functionality to cache these, or reading from the backend using selective queries? Not sure how much work the former creates.

Reply:
> instead of enforcing a maximum we limited each auth to some percentage of in flight updates

Do we have a count of auth servers available in the backend?

Reply:
@rosstimothy let me know if the new Scalability section addresses your questions.

Reply:
(No longer necessary in the current design.)